### PR TITLE
 ENH: add check_size option to ctypedef class for external classes 

### PR DESCRIPTION
**Pull request by mattip**
Address #2221 by adding a new `check_size` clause to `ctypedef class` definitions. This will allow suppressing the warning.

The srctree test actually comprises a few tests, but they are reported as one pass/fail. I am not sure how to break them into separate tests.

I think 'min', True, and False were the proposed values, but maybe I misunderstood.

### CODE REVIEW

**Comment by mattip** [comment]
Of course, we can fix the original issue by using this in `numpy.__init__.pxd`.

**Comment by mattip** [comment]
Yes, it seems filename.pxd creates a filename_api.h without the proto section, so it misses the declaration of the enum I tried to create.

File: Cython/Compiler/Parsing.py [hunk 1]
```diff
@@ -3521,13 +3522,15 @@ def p_c_class_definition(s, pos,  ctx):
...
         typeobj_name = typeobj_name,
+        check_size = check_size,
         in_pxd = ctx.level == 'module_pxd',
         doc = doc,
         body = body)
 
 def p_c_class_options(s):
     objstruct_name = None
     typeobj_name = None
+    check_size = b'min'
```
Comment (scoder) [inline Cython/Compiler/Parsing.py]: I'd like to have this a normal text string rather than a bytes string. Comparison to bytes might fail then parsed from Py3 code with unicode literals.
One way to do that might be to use the identifier `min` instead of the string literal `"min"`. Or check the literal node type and decode at need.

File: docs/src/userguide/extension_types.rst [hunk 2]
```diff
@@ -756,11 +764,23 @@ The part of the class declaration in square brackets is a special feature only
...
-    [object object_struct_name, type type_object_name ]
+    [object object_struct_name, type type_object_name, check_size cs_option]
+
+Where:
+
+- ``object_struct_name`` is the name to assume for the type's C struct.
+- ``type_object_name`` is the name to assume for the type's statically
+  declared type object.
+- ``cs_option`` is ``'min'`` (the default), ``True``, or ``False`` and is only
+  used for external extension types. If ``True``, ``sizeof(object_struct)`` must
```
Comment (scoder) [inline docs/src/userguide/extension_types.rst]: "only used for external extension types" - does the patch check that it's not used in other contexts? (And, are there actually other contexts?)

File: runtests.py [hunk 3]
```diff
@@ -1743,11 +1746,15 @@ def runTest(self):
...
+                cmd.append(command)
+                out.append(_out)
+                err.append(_err)
             res = p.returncode
             if res != 0:
-                sys.stderr.write("%s\n%s\n%s\n" % (
-                    command, self._try_decode(out), self._try_decode(err)))
+                for c, o, e in zip(cmd, out, err):
+                    sys.stderr.write("%s\n%s\n%s\n\n" % (
+                        c, self._try_decode(o), self._try_decode(e)))
```
Comment (scoder) [inline runtests.py]: This was already part of the other PR, right?

File: Cython/Compiler/ModuleNode.py [hunk 4]
```diff
@@ -3059,10 +3059,21 @@ def generate_type_import_call(self, type, code, import_generator, error_code=Non
         else:
             code.put('sizeof(%s), ' % objstruct)
 
-        code.putln('%i); if (!%s) %s' % (
-            not type.is_external or type.is_subclassed,
-            type.typeptr_cname,
-            error_code))
+        # check_size
+        if not type.is_external or type.is_subclassed:
```
Comment (scoder) [inline Cython/Compiler/ModuleNode.py]: Would be better to raise a compiler error here if `check_size` was set. Or at least warn that the value was ignored.

File: Cython/Utility/ImportExport.c [hunk 5]
```diff
@@ -308,15 +308,22 @@ static int __Pyx_SetPackagePathFromImportLib(const char* parent_package_name, Py
...
 #define __PYX_HAVE_RT_ImportType
 static PyTypeObject *__Pyx_ImportType(PyObject *module, const char *module_name, const char *class_name,
-    size_t size, int strict)
+    size_t size, int check_size)
 {
+    /*
+     * check_size tells what to do if tp_basicsize is different from size:
+     * 0 - Error (originates in check_size=True)
+     * 1 - Error if tp_basicsize is smaller, warn if larger (originates in check_size='min')
+     * 2 - Error if tp_basicsize is smaller (originates in check_size=False)
```
[the remark this answers is not in the document]
Comment (mattip) [reply Cython/Utility/ImportExport.c]: I think we always need to error if `tp_basicsize` is smaller than the C struct declared as the object, see the `_check_size1.pyx` test where I define a C struct that is bigger. Accessing `f2` would touch invalid memory.

So I claim even if using False we need to error on the side of caution.Open to other suggestions for the name.

File: docs/src/userguide/extension_types.rst [hunk 6]
```diff
@@ -736,6 +736,14 @@ built-in complex object.::
             ...
         } PyComplexObject;
 
+       At runtime, a check will be performed when importing the cython
+       c-extension module that ``__builtin__.complex``'s ``tp_basicsize``
+       matches ``sizeof(`PyComplexObject)``. This check can fail if in the cython
```
[the remark this answers is not in the document]
Comment (mattip) [reply docs/src/userguide/extension_types.rst]: ok

File: docs/src/userguide/extension_types.rst [hunk 7]
```diff
@@ -736,6 +736,14 @@ built-in complex object.::
             ...
         } PyComplexObject;
 
+       At runtime, a check will be performed when importing the cython
```
[the remark this answers is not in the document]
Comment (mattip) [reply docs/src/userguide/extension_types.rst]: ok

File: docs/src/userguide/extension_types.rst [hunk 8]
```diff
@@ -736,6 +736,14 @@ built-in complex object.::
...
+       c-extension module that ``__builtin__.complex``'s ``tp_basicsize``
+       matches ``sizeof(`PyComplexObject)``. This check can fail if in the cython
+       c-extension module was compiled with one version of the
+       ``complexobject.h`` header but imported into a python with a changed
```
[the remark this answers is not in the document]
Comment (mattip) [reply docs/src/userguide/extension_types.rst]: ok

File: runtests.py (hunk 3 above)
Comment (mattip) [reply runtests.py]: yes, sorry. Removing.

File: Cython/Compiler/ModuleNode.py (hunk 4 above)
Comment (mattip) [reply Cython/Compiler/ModuleNode.py]: I am not sure what these values mean. It is from the original code and the source of the documentation problem you pointed out.

File: docs/src/userguide/extension_types.rst (hunk 2 above)
Comment (mattip) [reply docs/src/userguide/extension_types.rst]: I was relating to the `if not type.is_external or type.is_subclassed` check done in `ModuleNode.py`, which I admit I don't really understand. See last comment below. A suggestion for rephrasing that better catches the logic is welcome

File: Cython/Compiler/ModuleNode.py (hunk 4 above)
Comment (mattip) [reply Cython/Compiler/ModuleNode.py]: Check added, test needed. Suggestion for `ctypedef` stanza that will trigger the failure welcome.

File: Cython/Compiler/Parsing.py (hunk 1 above)
Comment (mattip) [reply Cython/Compiler/Parsing.py]: it is messy since we want min, True, or False. Changing from p_atom(s) to p_ident(s) makes True and False into strings not booleans, is that OK? In any case, that enables removing the `b` (as long as we use a bare `min` not `'min'`. Changed documentation (which was missing quotes anyway) and test too.

**Comment by mattip** [comment]
Maybe worth adding some additional tests for failures and corner cases? Can we somehow write them as non-srctree tests?

File: Cython/Compiler/ModuleNode.py (hunk 4 above)
Comment (scoder) [reply Cython/Compiler/ModuleNode.py]: "is_external" = defined in a "cdef extern" block. False if a type is cimported from another module via a .pxd file.
"is_subclassed" = we found a subtype of the type, i.e. it is definitely not a leaf class of the hierarchy.

File: Cython/Compiler/Parsing.py [hunk 9]
```diff
@@ -3543,7 +3543,7 @@ def p_c_class_options(s):
             typeobj_name = p_ident(s)
         elif s.systring == 'check_size':
             s.next()
-            check_size = p_atom(s).value
+            check_size = p_ident(s)
```
Comment (scoder) [inline Cython/Compiler/Parsing.py]: We should check the valid values right here. No need to wait until code generation.
And if a simple adaptation or mapping of the values is needed, this can be done here, too.

**Comment by scoder** [comment]
> Can we somehow write them as non-srctree tests?

Don't think so. You need at least two files in order to cimport from the other.
You migth be able to cover multiple cases with the same srctree file, though. These tests are quite expensive to run, so it's not uncommon to cover an entire (little) feature with one test setup.

File: Cython/Compiler/Parsing.py (hunk 9 above)
Comment (mattip) [reply Cython/Compiler/Parsing.py]: done, should it be a function?

Comment (scoder) [reply Cython/Compiler/Parsing.py]: No, looks good to me as it is.

**Comment by mattip** [comment]
Thanks for merging. Now on to using this in the numpy pxd ...

**Comment by scoder** [comment]
Could you quickly write up a PR for that? I'd like to get the second RC out today.

**Comment by scoder** [comment]
Actually, I'll just enable ``check_size=False`` for the ndarray and dtype types, that should be it.

File: Cython/Utility/ImportExport.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (robertwb) [reply Cython/Utility/ImportExport.c]: If I understand the options correctly, we always error fo smaller structs, but either ignore, warn, or error for larger structs. This doesn't map obviously to the True/False/'min' options provided here--maybe literally call then None, 'warn,' and 'error'? (As an aside, maybe it'd be better to use an enum than 0, 1, 2 as well). 

(I actually wonder if instead we'd be better off doing a stricter check using offsetof for declared fields, even for non-extern classes, but that may have issues...)

File: Cython/Compiler/ModuleNode.py (hunk 4 above)
Comment (robertwb) [reply Cython/Compiler/ModuleNode.py]: Or do we want to allow this for non-extern classes? (We need to take care that the vtable doesn't change either...)

**Comment by mattip** [comment]
Thanks @scoder

File: Cython/Utility/ImportExport.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (scoder) [reply Cython/Utility/ImportExport.c]: What about "extend", "warn" and "error" ? `None` doesn't tell me much either.

Agreed that offset checks would be nice, but that's a separate feature (and not for 0.29).

**Comment by scoder** [comment]
I changed the available options to "extend", "warn" and "error". The boolean values really were everything but obvious.

File: Cython/Compiler/ModuleNode.py (hunk 4 above)
Comment (mattip) [reply Cython/Compiler/ModuleNode.py]: If the type is `cimport`ed, how would we set `check_size`? Is there an example in the tests that redefines a `cimport`ed type?

**Comment by robertwb** [comment]
Perhaps the options should be "ignore", "error", and "warn"?

Not sure why putting an enum right before the __Pyx_ImportType declaration
didn't work, that's what I would have done.

As for non-extern classes, it would help if one compiled against an an
older version of a pxd. One would still have to ensure the vtables are
compatible, and of course it would have to be final. Just a half-baked
thought there, something we could always flesh out and enable later if
there was a need.

Speaking of which, are we marking these no-size-check classes as final?

**Comment by scoder** [comment]
>Perhaps the options should be "ignore", "error", and "warn"?

Works for me, feel free to change it. I already reduced the number of places that need to be adapted.
